crates/wasmparser/src/validator/operators.rs

```diff
@@ -626,8 +638,15 @@ impl OperatorValidator {
         &mut self,
         resources: &impl WasmModuleResources,
     ) -> OperatorValidatorResult<()> {
-        for ty in results(self.control[0].block_type, resources)?.rev() {
-            self.pop_operand(Some(ty), resources)?;
+        for expected in results(self.control[0].block_type, resources)?.rev() {
+            let actual = self.pop_operand(Some(expected), resources)?;
+            if !resources.matches(actual, expected) {
+                bail_op_err!(
+                    "type mismatch: expected {}, found {}",
+                    ty_to_str(expected),
+                    ty_to_str(actual)
+                );
+            }
```

Comment:
This is certainly redundant, right?

Reply:
Possibly! I will have a look at it again.